koku/api/report/queries.py

```diff
@@ -315,67 +323,76 @@ def _get_search_filter(self, filters):  # noqa C901
         LOG.debug(f"self.query_exclusions: {self.query_exclusions}")
         return composed_filters
 
-    def _set_or_filters(self):
+    def _set_or_filters(self, or_filter=None):
         """Create a composed filter collection of ORed filters.
 
         This is designed to handle specific cases in the provider_map
         not to accomodate user input via the API.
 
         """
         filters = QueryFilterCollection()
-        or_filter = self._mapper._report_type_map.get("or_filter", [])
+        if not or_filter:
+            or_filter = self._mapper._report_type_map.get("or_filter", [])
         for filt in or_filter:
             q_filter = QueryFilter(**filt)
             filters.add(q_filter)
 
         return filters.compose(logical_operator="or")
 
-    def _set_tag_filters(self, filters, check_for_exclude=False):
+    def _set_tag_exclusion_filters(self):
+        """Creates exclusion fitlers for tags that allow null returns."""
+        tag_exclusion_collection = QueryFilterCollection()
+        emptyset_filters = []
+        tag_filters = self.get_tag_filter_keys(parameter_key="exclude")
+        for tag in tag_filters:
+            tag_db_name = self._mapper.tag_column + "__" + strip_tag_prefix(tag)
+            filt = {"field": tag_db_name, "operation": "notlikelist"}
+            emptyset_filters.append({"field": tag_db_name, "operation": "isnull", "parameter": True})
+            list_ = self.parameters.get_exclude(tag, list())
+            if list_ and not ReportQueryHandler.has_wildcard(list_):
+                q_filter = QueryFilter(parameter=list_, **filt)
+                tag_exclusion_collection.add(q_filter)
+        emptyset_filters.append({"field": self._mapper.tag_column, "operation": "exact", "parameter": "{}"})
+        emptyset_composed = self._set_or_filters(emptyset_filters)
```

Comment:
How does an `or` filter behave with exclusions?

Reply:
So, these `or_filters` are to add additional filters I needed to include the emptysets to get `no-option` to show up. So, it will add filters like `tag is null or tag = '{}'` to the sql query. However, in general, I think this could be a worthwhile discussion to see if or is needed with excludes.

Reply:
So, I continued digging into this question, and with the `noticontainslist` filter `exclude[or:tag:app]` would do the same thing as `exclude[and:tag:app]` because of the logic I added [here](https://github.com/project-koku/koku/pull/3997/files#diff-85e41f2ec3229710ac79109eb26a6463faf136a8b9acdd677d57a5b34798a6fcR29). If we change that logical operator to an `OR` than when we do multiple excludes:
```
((UPPER((reporting_awscostentrylineitem_daily_summary.tags ->> 'app')::text) NOT LIKE UPPER('%analytics%')) OR (UPPER((reporting_awscostentrylineitem_daily_summary.tags ->> 'app')::text) NOT LIKE UPPER('%catalog%'))
```
both the `analytics` & `catalog` tag will be returned in the response. So, the short answer here for tags is that it is essentially doesn't do anything. I can update the serializer to not allow `exclude[or:tag:{key}]` or we can just keep it & do the right thing on the backend. 

-------
**Outside of Tag Exclusions**

`exclude[or:{option}]={value}` is needed in order to tell django how to combine two different exclusions. For example, if you were to hit the following endpoint:
```
http://localhost:8000/api/cost-management/v1/reports/aws/costs/?group_by[account]=*&group_by[service]=*&exclude[account]=9999999999995&exclude[service]=AmazonRDS
```
Django creates sql like this:
```
AND NOT (((UPPER(reporting_awsaccountalias.account_alias::text) LIKE UPPER('%9999999999995%') 
            AND reporting_awsaccountalias.account_alias IS NOT NULL) 
            OR UPPER(reporting_aws_cost_summary_by_service_p.usage_account_id::text) LIKE UPPER('%9999999999995%')) 
            AND UPPER(reporting_aws_cost_summary_by_service_p.product_code::text) LIKE UPPER('%AmazonRDS%'))
```
Due to the `AND` this will only exclude the `AmazonRDS` when the accounts = `9999999999995` which we are about to exclude with the `NOT` making that condition useless.

However, if we change it to: `
&exclude[account]=9999999999995&exclude[or:service]=AmazonRDS`
```
AND NOT (
        ((UPPER(reporting_awsaccountalias.account_alias::text) LIKE UPPER('%9999999999995%') 
        AND reporting_awsaccountalias.account_alias IS NOT NULL)  
        OR UPPER(reporting_awscostentrylineitem_daily_summary.usage_account_id::text) LIKE UPPER('%9999999999995%') 
        OR UPPER(reporting_awscostentrylineitem_daily_summary.product_code::text) LIKE UPPER('%AmazonRDS%')))) 
```
That `AND` is flipped to an `OR` allowing us to exclude `AmazonRDS` from the return for all the accounts not just the one we excluded.

Reply:
**Additional Context:**
So, Michael Andrew and I had a private chat in slack regarding:
```
exclude[account]=9999999999995&exclude[service]=AmazonRDS
Does this query make sense to only exclude AmazonRDS when the account equals 9999999999995? I feel like if I was making that API request, i would want to exclude everything that is account 9999999999995 and every service that is AmazonRDS
```
After discussion we agreed that the logic is confusing, and we were making:
```
WHERE NOT (account = '9999999999995' AND service = 'AmazonRDS')
```
instead of
```
WHERE NOT account = '9999999999995' AND NOT service = 'AmazonRDS'
```
I attempted to configure django to do `WHERE NOT account = '9999999999995' AND NOT service = 'AmazonRDS'`. However, I had some issues with how django to do this. However, I was able to get it to do:
```
WHERE NOT (account = '9999999999995' OR service = 'AmazonRDS')
```

---- 

Additional query checks to make sure they were equivalent:
NOT FOO and NOT bar:
```
SELECT
    sum(reporting_aws_cost_summary_by_service_p.unblended_cost),
    reporting_aws_cost_summary_by_service_p.product_code AS service
FROM reporting_aws_cost_summary_by_service_p
LEFT OUTER JOIN reporting_awsaccountalias ON (reporting_aws_cost_summary_by_service_p.account_alias_id = reporting_awsaccountalias.id)
WHERE (
    reporting_aws_cost_summary_by_service_p.usage_start = '2022-11-07'::date
    AND NOT (((UPPER(reporting_awsaccountalias.account_alias::text) LIKE UPPER('%9999999999995%') AND reporting_awsaccountalias.account_alias IS NOT NULL)))
    AND NOT (UPPER(reporting_aws_cost_summary_by_service_p.usage_account_id::text) LIKE UPPER('%9999999999995%')))
    GROUP BY DATE_TRUNC('day', reporting_aws_cost_summary_by_service_p.usage_start), reporting_aws_cost_summary_by_service_p.product_code, COALESCE('USD', 'USD');
```
result:
```
      sum      |             service
---------------+----------------------------------
   1.000028000 | AmazonRoute53
  22.633787904 | AmazonRDS
  12.000000000 | AmazonS3
 454.259754079 | Red Hat OpenShift Service on AWS
   0.469870676 | AmazonDynamoDB
  39.625239100 | AmazonEC2
   1.200000000 | AmazonVPC
```
NOT (FOO OR BAR)
```
SELECT
    sum(reporting_aws_cost_summary_by_service_p.unblended_cost),
    reporting_aws_cost_summary_by_service_p.product_code AS service
FROM reporting_aws_cost_summary_by_service_p
LEFT OUTER JOIN reporting_awsaccountalias ON (reporting_aws_cost_summary_by_service_p.account_alias_id = reporting_awsaccountalias.id)
WHERE (
    reporting_aws_cost_summary_by_service_p.usage_start = '2022-11-07'::date
    AND
    NOT (((UPPER(reporting_awsaccountalias.account_alias::text) LIKE UPPER('%9999999999995%') AND reporting_awsaccountalias.account_alias IS NOT NULL)
            OR UPPER(reporting_aws_cost_summary_by_service_p.usage_account_id::text) LIKE UPPER('%9999999999995%')
        ))) GROUP BY DATE_TRUNC('day', reporting_aws_cost_summary_by_service_p.usage_start), reporting_aws_cost_summary_by_service_p.product_code, COALESCE('USD', 'USD');
```
result:
```
      sum      |             service
---------------+----------------------------------
   1.000028000 | AmazonRoute53
  22.633787904 | AmazonRDS
  12.000000000 | AmazonS3
 454.259754079 | Red Hat OpenShift Service on AWS
   0.469870676 | AmazonDynamoDB
  39.625239100 | AmazonEC2
   1.200000000 | AmazonVPC
```